mlflow/projects/utils.py

```diff
@@ -80,3 +100,186 @@ def _is_valid_branch_name(work_dir, version):
+    # executable is availble on the PATH, so we only want to fail if we actually need it.
+    import git
+    repo = git.Repo.init(dst_dir)
+    origin = repo.create_remote("origin", uri)
+    origin.fetch()
+    if version is not None:
+        try:
+            repo.git.checkout(version)
+        except git.exc.GitCommandError as e:
+            raise ExecutionException("Unable to checkout version '%s' of git repo %s"
+                                     "- please ensure that the version exists in the repo. "
+                                     "Error: %s" % (version, uri, e))
+    else:
+        repo.create_head("master", origin.refs.master)
+        repo.heads.master.checkout()
+    repo.submodule_update(init=True, recursive=True)
+
+
+def _fetch_zip_repo(uri):
+    import requests
+    from io import BytesIO
+    # TODO (dbczumar): Replace HTTP resolution via ``requests.get`` with an invocation of
+    # ```mlflow.data.download_uri()`` when the API supports the same set of available stores as
+    # the artifact repository (Azure, FTP, etc). See the following issue:
+    # https://github.com/mlflow/mlflow/issues/763.
+    response = requests.get(uri)
+    try:
+        response.raise_for_status()
+    except requests.HTTPError as error:
+        raise ExecutionException("Unable to retrieve ZIP file. Reason: %s" % str(error))
+    return BytesIO(response.content)
+
+
+def get_or_create_run(run_id, uri, experiment_id, work_dir, entry_point):
+    if run_id:
+        return tracking.MlflowClient().get_run(run_id)
+    else:
+        return _create_run(uri, experiment_id, work_dir, entry_point)
+
+
+def _create_run(uri, experiment_id, work_dir, entry_point):
+    """
+    Create a ``Run`` against the current MLflow tracking server, logging metadata (e.g. the URI,
+    entry point, and parameters of the project) about the run. Return an ``ActiveRun`` that can be
+    used to report additional data about the run (metrics/params) to the tracking server.
+    """
+    if _is_local_uri(uri):
+        source_name = tracking._tracking_service.utils._get_git_url_if_present(_expand_uri(uri))
+    else:
+        source_name = _expand_uri(uri)
+    source_version = _get_git_commit(work_dir)
+    existing_run = fluent.active_run()
+    if existing_run:
+        parent_run_id = existing_run.info.run_id
+    else:
+        parent_run_id = None
+
+    tags = {
+        MLFLOW_USER: _get_user(),
+        MLFLOW_SOURCE_NAME: source_name,
+        MLFLOW_SOURCE_TYPE: SourceType.to_string(SourceType.PROJECT),
+        MLFLOW_PROJECT_ENTRY_POINT: entry_point
+    }
+    if source_version is not None:
+        tags[MLFLOW_GIT_COMMIT] = source_version
+    if parent_run_id is not None:
+        tags[MLFLOW_PARENT_RUN_ID] = parent_run_id
+
+    active_run = tracking.MlflowClient().create_run(experiment_id=experiment_id, tags=tags)
+    return active_run
+
+
+def log_project_params_and_tags(run, project, work_dir, entry_point, parameters, version):
```

Comment:
Should this be internal (`_log_project_params_and_tags`) if the primary API we expect plugin developers to call is `prepare_project_and_run`?

Reply:
As the `prepare_project_and_run` is removed (to take in account your other comment). We should now keep them public as they will call by backend developers.